File: ROS/C++/src/talker.cpp
#include "ros/ros.h"

#include "sensor_msgs/Image.h"


#include <chrono>

int main(int argc, char **argv)
{

  ros::init(argc, argv, "talker");

  ros::NodeHandle n;

  ros::Publisher chatter_pub = n.advertise<sensor_msgs::Image>("chatter", 1000);

  ros::Rate loop_rate(20);

  static double logmul[] = {2.0, 2.5, 2.0};
  int datasize = 100;
  const int maxsize = 1200001;
  int count = 0;
  const int datacount = 500;
  while (ros::ok() && datasize < maxsize)
  {
    
    sensor_msgs::Image msg;

    //std::stringstream ss;
    //ss << "hello world " << count;
    //msg.data = ss.str();

    //ROS_INFO("%s", msg.data.c_str());

    msg.height = 100;
    msg.width = datasize/msg.height/3;
    msg.data.resize(msg.height*msg.width*3);
    auto now = std::chrono::system_clock::now().time_since_epoch();
    auto sec = std::chrono::duration_cast<std::chrono::seconds>(now);
    auto nsec = std::chrono::duration_cast<std::chrono::nanoseconds>(now - sec);
    msg.header.stamp.sec  = static_cast<unsigned int>(sec.count());
    msg.header.stamp.nsec = static_cast<unsigned int>(nsec.count());


    chatter_pub.publish(msg);

    


    loop_rate.sleep();
    //if(count%30 == 29)
    // datasize += 5000;
    ++count;
    
    if(count%datacount == 0)
    {
      std::cout << msg.height*msg.width*3 << std::endl;
      datasize = static_cast<int>(static_cast<double>(datasize)*logmul[(count/datacount)%3]);
    }
    
  }
  std::cout << "finish" << std::endl;
  while (ros::ok())
  {
    loop_rate.sleep();
  }


  return 0;
}

File: ROS/Python/src/talker.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# -*- Python -*-


import rospy
from sensor_msgs.msg import Image
import time



def talker():
    pub = rospy.Publisher('chatter', Image, queue_size=10)
    rospy.init_node('talker', anonymous=True)
    rate = rospy.Rate(500)

    datasize = 100
    count = 0

    while not rospy.is_shutdown():
        msg = Image()
        maxsize = 12000001
        datacount = 50

        logmul = [2.0, 2.5, 2.0]

        if datasize > maxsize:
        	print("finish")
        	return

        msg.height = 100
        msg.width = datasize/msg.height/3
        msg.encoding = "rgb8"
        msg.step = 1920
        msg.data = " "*(msg.height*msg.width*3)

        
        #rospy.loginfo(hello_str)
        
        tm = time.time()
        tm_f = float(tm - long(tm))
        msg.header.stamp.secs = long(float(tm) - float(tm_f))
        msg.header.stamp.nsecs = long(float(tm_f) * float(1000000000))

        pub.publish(msg)

        count += 1
        if count%datacount == 0:
        	print(msg.height*msg.width*3)
        	datasize = int(datasize*logmul[(count/datacount)%3])

        rate.sleep()

if __name__ == '__main__':
    try:
        talker()
    except rospy.ROSInterruptException:
        pass



    